api/controllers/clients.js

```diff
@@ -0,0 +1,224 @@
+
+async function getClientByEmail (email) {
+    const result = await db.query(`SELECT * FROM ${TABLE} WHERE ${EMAIL} = '${email}';`);
+    return result.rows.length > 0 ? result.rows[0] : null;
+}
+
+function checkForEmailAndPassword (request, action) {
+    let email = assertHeaderKey (request, EMAIL, action);
+    let password = assertHeaderKey (request, PASSWORD, action);    
+    return { email, password };
+}
+
+async function assertClientAuth (request, action) {
+    let { email, password } = checkForEmailAndPassword (request, action);
+    function throwAuthError () {
+        throw new Error(`${action} :: Authorization Failed [Invalid Email Or Password]!`);
+    }
+    let client = await getClientByEmail(email);
+    if (!client) 
+        throwAuthError ();
+    if (!(await bcrypt.compare(password, client[PASSWORD])))
+        throwAuthError ();
+    return client;
+}
+
+
+/*
+INPUT: { NONE }
+OUTPUT
+{
+    clients: [ 
+        { < client object with info > },
+        { < client object with info > },
+        ...
+    ]
+}
+*/
+module.exports.get_clients_list = async (req, response, next) => {
+    try {
+        let clients = (await db.query(`SELECT ${CLIENT_ID}, ${DISPLAY_NAME}, ${BIO} FROM ${TABLE} WHERE ${CLIENT_ID} != '${SAFETRACE_CLIENT_ID}';`)).rows;
+        response.status(200).json( { clients: clients } );
+    }
+    catch (e) { next(e); }
+};
+
+
+/*
+INPUT:
+{
+    HEADER
+    {
+        email:      < email >,
+        password:   < password >,
+    }
+    display_name:   < display name >
+    bio:            < a short bio >
+}
+OUTPUT
+{
+    message:        `Account Created For: < email >, save the API and Private Keys included in this object. Make sure to keep them private and secure`,
+    api_key:        < api key >,
+    private_key:    < private key >
+    public_key:     < public key >
+}
+*/
+module.exports.client_signup = async (req, response, next) => {
+    try {
+        const action = 'SIGN UP';
+        let { email, password } = checkForEmailAndPassword (req, action);
+        
+        let bio = assertBodyKey (req, BIO, action);
+        let displayName = assertBodyKey (req, DISPLAY_NAME, action);
+        
+        function throwError () {
+            throw new Error(`${action} :: Invalid Or Existing Email Address: ${email}`);
+        }
+        
+        if (!validEmailRegex.test(email))
+            throwError();
+        
+        if ((await getClientByEmail(email))) 
+            throwError();
+            
+        const passwordHash = await hashing.hashStringRandom(password);
+            
+        // make uuid for api key
+        let apiKey = await v1(); 
+        let encKeys = a_Encryption.getKeyPair();
+
+        let encryptedAPIKey = s_Encryption.encryptData(apiKey, process.env.SAFETRACE_KEY, process.env.SAFETRACE_IV);
```

Comment:
is SAFETRACE_IV static? IV doesn't have to be secret, but it does have to be unique for each encryption.

Reply:
SAFETRACE_IV is static, I just need it to be consistent since I've been checking the API key be re-encrypting it to compare to the encrypted version in the DB

Reply:
So IV should not be static but needs to be randomly generated